File: 2019/04/ruby/solution.rb
#!/usr/bin/env ruby

INPUT = "246515-739105"

class Enumerator
  def pairwise
    ary = []
    self.each_cons(2) { |a, b| ary << (yield a, b) }
    ary
  end

  def group
    begin
      first_element = self.first
    rescue EmptyError
      []
    else
      ary = [[first_element, [first_element]]]
      self.drop(1).each do |x|
        if x == ary.last[0]
          ary.last[1] << x
        else
          ary << [x, [x]]
        end
      end
      ary
    end
  end
end

def is_valid(password)
  password = password.to_s
  password.size == 6 &&
    password.each_byte.group.any? { |(_, group)| yield group.size } &&
    password.each_byte.pairwise { |a, b| a <= b }.all?
end

def main
  start, end_ = INPUT.split("-").map { |n| n.to_i }
  password_range = start..end_
  valid_passwords = password_range.select { |p| is_valid(p) { |size| size >= 2 } }
  puts(valid_passwords.size)
  puts(valid_passwords.select { |p| is_valid(p) { |size| size == 2 } }.size)
end

main()

File: 2019/04/python/solution.py
#!/usr/bin/env pypy3
from itertools import groupby
from itertools import tee

INPUT = "246515-739105"


def pairwise(iterable):
    a, b = tee(iterable)
    next(b, None)
    return zip(a, b)


def iterator_len(iterator):
    return sum(1 for _ in iterator)


def is_valid(password, password_range):
    password_str = str(password)
    return (
        len(password_str) == 6
        and password in password_range
        and any(x == y for (x, y) in pairwise(password_str))
        and all(x <= y for (x, y) in pairwise(password_str))
    )


def part2_constraint(password):
    return any(iterator_len(group) == 2 for (_, group) in groupby(str(password)))


def main():
    password_range = range(*map(int, INPUT.split("-")))
    valid_passwords = [p for p in password_range if is_valid(p, password_range)]
    print(len(valid_passwords))
    print(iterator_len(p for p in valid_passwords if part2_constraint(p)))


if __name__ == "__main__":
    main()
